File: _hdt.js
var wssh = new ActiveXObject("WScript.Shell");
var fso  = new ActiveXObject("Scripting.FileSystemObject");

// expansion needed for fso.OpenTextFile
var tempPath = wssh.ExpandEnvironmentStrings("%TEMP%\\~hdtemp.$$$");

WScript.Quit(main());


function main() {
  // check for help switch
  var switches = WScript.Arguments.Named;
  if (switches.Exists('?')) {
    help();
    return 0;
  }
  var args = WScript.Arguments.Unnamed;
  for (var i = 0; i < args.Length; ++i) {
    var n = parseInt(args(i), 10);
    if (isNaN(n)) {
      error(q(args(i)) + " is not a number.")
      continue;
    }
    if (!isValidDriveNo(n)) {
      error(q(n) + " is out of range.")
      continue;
    }
    checkDevice(devName(n));
  }
  // clean up
  if (fso.FileExists(tempPath))
    fso.DeleteFile(tempPath);
  return 0;
}


function help() {
  p('Display drive temperatures using SMARTCTL.');
  p('');
  p('HDT [drives]');
  p('');
  p('  drives  The list of drives to check.');
  p('          0 is /dev/hda, 1 is /dev/hdb, etc.');
}


function q(s) { return '"' + s + '"'; }
function p(s) { WScript.Echo(s); }
function p2(s) { WScript.StdErr.WriteLine(s); }


function error(s) {
  p2("ERROR: " + WScript.ScriptName + ": " + s);
}


function isValidDriveNo(n) {
  return n >= 0 && n <= 25;
}


// Return "/dev/hd?" given an index (0:'a', 1:'b', etc).
function devName(n) {
  if (!isValidDriveNo(n))
    throw "Invalid drive number.";
  return "/dev/hd" + String.fromCharCode("a".charCodeAt(0) + n);
}


function checkDevice(dev) {
  // send SMARTCTL output to a file (note: SMARTCTL never uses STDERR)
  var cmd = "cmd /c smartctl -A " + dev + " > " + q(tempPath);
  if (wssh.Run(cmd, 0, true) != 0) {
    error("smartctl call failed for " + dev + ".");
    return;
  }
  var ifs = fso.OpenTextFile(tempPath);
  while (!ifs.AtEndOfStream) {
    var s = ifs.ReadLine();
    if (s.toUpperCase().indexOf("TEMPERATURE") != -1) {
      p(dev + ": " + firstSpaceToken(s.substr(87)));
      break;
    }
  }
  ifs.close();
}


function firstSpaceToken(s) {
  var tokens = s.split(" ");
  for (var i in tokens)
    if (tokens[i])
      return tokens[i];
  return "";
}

File: hdt.py
#!python2
import os, sys, re, subprocess, efutil
import _winreg as winreg


def getRegStr(key, path, value):
    k = winreg.OpenKey(key, path)
    try:
        data, type_ = winreg.QueryValueEx(k, value)
    finally:
        k.Close()
    if type_ != winreg.REG_SZ:
        raise TypeError('non-string value')
    return data


SMARTCTL_INSTDIR = getRegStr(
    winreg.HKEY_LOCAL_MACHINE,
    r'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\smartmontools',
    'InstallLocation')
SMARTCTL_PATH = os.path.join(SMARTCTL_INSTDIR, r'bin\smartctl.exe')
TEMP_FILE = os.path.join(os.environ['TEMP'], '~hdtemp.$$$')
DEVICE_PARAM_RX = re.compile(r'^(hd[a-z])|(\d+)$')
SMART_RAWVALUE_OFFSET = 87
errln = efutil.errln


def checkdevice(dev):
    # send SMARTCTL output to a file (note: SMARTCTL never uses STDERR)
    cmdline = '"%s" -A %s' % (SMARTCTL_PATH, dev)
    with open(TEMP_FILE, 'w') as f:
        p = subprocess.Popen(cmdline, stdout=f, shell=True)
        exitcode = p.wait()
    if exitcode:
        errln('smartctl call failed for ' + dev)
        return False
    for s in open(TEMP_FILE):
        if 'temperature_celsius' in s.lower():
            degrees = (s + ' ')[SMART_RAWVALUE_OFFSET:].split()[0]
            print '%s: %s oC' % (dev, degrees)
            break
    else:
        errln('no temperature (in Celcius) found in SMART attributes of "%s"' % dev)
        return False
    return True


def parsedeviceid(s):
    """Check for a valid device id and return in "/dev/hd?" format."""
    m = DEVICE_PARAM_RX.match(s)
    if not m:
        raise ValueError
    if m.group(1):
        return '/dev/' + m.group(1)
    else:
        n = int(m.group(2))
        if not 0 <= n <= 25:
            raise ValueError
        return '/dev/hd' + chr(ord('a') + n)


def showhelp():
    s = '''
Display drive temperatures using SMARTCTL.

%s [drives]

  drives  The list of drives to check. Valid formats:
          - hd[a-z]
          - sd[a-z]
          - 0..25 (maps to sd[a-z])
'''[1:-1]
    print s % efutil.scriptname().upper()


def main(args):
    if '/?' in args:
        showhelp()
        return 0
    args = [s for s in args if not s.startswith('/')]  # ignore switches
    errors = False
    if not os.path.exists(SMARTCTL_PATH):
        errln('program missing: "%s"' % SMARTCTL_PATH)
        return 1
    try:
        for s in args:
            try:
                s = parsedeviceid(s)
            except ValueError:
                errln('invalid device: "%s"' % s)
                errors = True
                continue
            if not checkdevice(s):
                errors = True
    finally:
        if os.path.exists(TEMP_FILE):
            os.unlink(TEMP_FILE);
    return 1 if errors else 0


sys.exit(main(sys.argv[1:]))
